### PR TITLE
Fix bug when BSI uses min max values for evaluation and some other improvements

### PR DESCRIPTION
**Pull request by zhongyujiang**
### SUMMARY
- Describe your changes, including rationale and design decisions

This PR fixes the bug when BSI using the max value for evaluation, currently BSI uses below logic  in comparing: https://github.com/RoaringBitmap/RoaringBitmap/blob/d9d0914ea78d0b25fb5813159a0a1230d701b5ee/bsi/src/main/java/org/roaringbitmap/bsi/RoaringBitmapSliceIndex.java#L410-L415

which is not correct. For example, when the operation is `NEQ` or `LE`,  if the `startOrValue` is greater than the `maxValue`, the return should not be an empty set, but the full set of all values.
To fix this, this add a `compareUsingMinMax` method to deal with different operations. This also fixes the bug that the min value and the max value are not updated after the `add` operation,  this recalculate and update min and max values after each `add` call.  This PR adds more tests to cover evaluation of boundary values.
   
- If your pull request includes new functionality, then you must have corresponding new tests.
This PR does not introduce new functionalities.
- We require all Java files to be ASCII-only to avoid compatibility problems.

### Automated Checks

- [x] I have run `./gradlew test` and made sure that my PR does not break any unit test.
- [x] I have run `./gradlew checkstyleMain` or the equivalent and corrected the formatting warnings reported.

### CODE REVIEW

File: bsi/src/main/java/org/roaringbitmap/bsi/BitmapSliceIndex.java [hunk 1]
```diff
@@ -75,8 +75,14 @@ enum Operation {
    * currentMaxValue/currentMinValue are optional,it's can be compute from input value list.
    * and avoiding bsi expend slice array capacity.
    */
+  @Deprecated
   void setValues(List<Pair<Integer, Integer>> values, Integer currentMaxValue, Integer currentMinValue);
```
Comment (zhongyujiang) [inline bsi/src/main/java/org/roaringbitmap/bsi/BitmapSliceIndex.java:79]: Deprecated this as I think it's more natural to calculate the min-max from the actual values, rather than setting them externally, such operations should be handed over to a capacity expansion interface.

**Comment by zhongyujiang** [comment]
@lemire Hey, can you please help review this? Thanks.

**Comment by lemire** [comment]
@zhongyujiang We are running the tests.

The code looks good to me.

**Comment by lemire** [comment]
Merging.
